[#13894] [Docs] Add docs page for concurrency control + miscellaneous related changes (#14839)
yugabyte/yugabyte-db: docs/content/preview/architecture/transactions/concurrency-control.md

---
title: Concurrency control
headerTitle: Concurrency control
linkTitle: Concurrency control
description: Details about Concurrency control in YSQL
menu:
  preview:
    identifier: architecture-concurrency-control
    parent: architecture-acid-transactions
    weight: 1153
type: docs
---

> [Concurrency control](https://en.wikipedia.org/wiki/Concurrency_control) in databases ensures that multiple transactions can execute concurrently while preserving data integrity. Concurrency control is essential for correctness in environments where two or more transactions can access the same data at the same time.

Author: I think "ensures" is the right word? Since "concurrency control" is singular?

> [Concurrency control](https://en.wikipedia.org/wiki/Concurrency_control) in databases ensures that multiple transactions can execute concurrently while preserving data integrity. Concurrency control is essential for correctness in environments where two or more transactions can access the same data at the same time.

Reviewer: oh my bad ! Sorry yes you're right :)

> YugabyteDB provides 2 policies to handle conflicts between concurrent transactions as described below. Also, see section at the end for information on how row-level explicit locking clauses behave with these concurrency control policies.

Reviewer: ```suggestion
YugabyteDB provides two policies to handle conflicts between concurrent transactions as described in the following sections. 

For information on how row-level explicit locking clauses behave with these concurrency control policies, refer to [Row-level explicit locking clauses](#row-level-explicit-locking-clauses) .
```

## Fail-on-Conflict

This is the default concurrency control strategy and is applicable for `REPEATABLE READ` and `SERIALIZABLE` isolation levels.
It is not applicable for [Read Committed](../read-committed/) isolation.

> 1. In this mode, transactions are assigned random priorities with some exceptions as described in [Transaction Priorities](../transaction-priorities/).

Reviewer: ```suggestion
In this mode, transactions are assigned random priorities with some exceptions as described in [Transaction Priorities](../transaction-priorities/).
```

> 2. If a conflict occurs, the transaction with lower priority is aborted. There are two possibilities when a transaction T1 tries to read/ write/ lock a row in a mode conflicting with a other concurrent transactions:

Reviewer: ```suggestion
If a conflict occurs, a transaction with the lower priority is aborted. There are two possibilities when a transaction T1 tries to read or write or lock a row in a mode conflicting with a other concurrent transactions as follows:
```

   - **Wound:** If T1 has a priority higher than all the other conflicting transactions, T1 will abort them and make progress.
   - **Die:** If any other conflicting transaction has an equal or higher priority than T1, T1 will abort itself.

> ```sql

Reviewer: ```suggestion

### Examples 

Start by creating a table and insert data to it using the following commands:
```sql
```

Author: Let us not say "Examples" maybe? Since this is to explain the "Wound" and "Die" approaches

Reviewer: Yes , in that case we can just have a statement before creating the table commands. Something like - "Suppose you have a table with some data in it, the following section describes the wound and die approaches when a conflict occurs".

Author: yes that is better, added it

create table test (k int primary key, v int);
insert into test values (1, 1);
```

### Wound example

<table class="no-alter-colors">
<thead>
    <tr>
    <th>
    Client 1
    </th>
    <th>
    Client 2
    </th>
    </tr>
  </thead>
  <tbody>
  <tr>
   <td>
   </td>
   <td>

```sql
SET yb_transaction_priority_upper_bound = 0.4;
```

> 

Author: Done, btw I realized that this happens only for the sql statements that spill into more than 1 line on the UI. Maybe we can fix this in our generic styling for future and so that it works well on small screens too

Each retry will work a newer snapshot of the database in anticipation that the conflicts might not occur (if an earlier conflicting transaction T2 has committed with a commit time before the read time of the new snapshot, the conflicts with T2 would essentially be voided since T1 and T2 would no longer be "concurrent").

Also note that the retries will not be performed in case the amount of data to be sent from YSQL to the client proxy exceeds the TServer gflag `ysql_output_buffer_size`.

## Wait-on-Conflict [[Beta]](/preview/faq/general/#what-is-the-definition-of-the-beta-feature-tag)

This mode of concurrency control is applicable only for YSQL and provides the same semantics as PostgreSQL.

> In this mode, transactions are not assigned priorities. If a conflict occurs when a transaction T1 tries to read/ write/ lock a row in a conflicting mode with a few other concurrent transactions, T1 will **wait** until all conflicting transactions finish by either committing or rolling back. Once all conflicting transactions have finished, T1 will:

Author: Changed to `read, write or lock`

Also, since there are just two instances of `read, write, or lock`, avoiding defining access for now. Let me know if I should still define and use `access` instead

1. make progress if the conflicting transactions didn’t commit any permanent modifications that conflict with T1.
2. abort otherwise.

By default, transactions operate in the priority based `Fail-on-Conflict` mode. This mode can be enabled by setting the YB-TServer gflag `enable_wait_queues=true`, which start in-memory wait queues that provide waiting semantics when conflicts are detected between transactions. A rolling restart is needed for the gflag to take effect.

Since T1 can make progress only if the conflicting transactions didn’t commit any conflicting permanent modifications, there are some intricacies in the behaviour. The list of exhaustive cases possible are detailed below in the Examples section.

{{< note >}}

> 

Author: done

1. Semantics of [Read Committed](../read-committed/) isolation make sense only with the Wait-on-Conflict behaviour.
1. Transactions in Read Committed isolation follow `Wait-on-Conflict` policy even without wait queues i.e., even when `enable_wait_queues=false`. This is done by relying on a indefinite retry-backoff mechanism with exponential delays when conflicts are detected. However, when `Read Committed` isolation provides Wait-on-Conflict semantics without wait queues, the following limitations exist -
    1. there might be a need to manually tune the exponential backoff parameters for performance as explained [here](../read-committed/#performance-tuning).
    1. the app will have to [rely on statement timeouts to avoid deadlocks](../read-committed/#avoid-deadlocks-in-read-committed-transactions).
    1. there can be unfairness and high P99 latencies during contention due to the retry-backoff mechanism.
{{</note >}}

> {{< note title="Best-effort internal retries for first statement in a transaction apply even in Wait-on-Conflict policy" >}}

Author: Since there are two suggestions, changed to this sort of middle path - `Best-effort internal retries also apply for the first statement in transactions following the Wait-on-Conflict policy`, let me know

> The best-effort internal retries described in Fail-on-Conflict apply even here. This is an extra enhancement which Postgres doesn't provide. After a transaction T1, that was waiting on other transactions, unblocks, it could be the case that some conflicting modifications were committed to the database. In this case T1 has to abort. However, if it was still the first statement that was being executed in T1, best-effort internal retries using a later snapshot of the database will be performed to possibly make progress.

Author: done

{{</note >}}

### Examples

> 

Author: done

> Set the TServer gflag `enable_wait_queues=true`. Also, set TServer gflag `ysql_max_write_restart_attempts=0` to avoid effects from best-effort internal retries for first statement in transaction block if it faces conflicts. A restart is necessary for these flags to take effect.

Author: Changed to - `Note that the examples require you to set the YB-TServer gflag `enable_wait_queues=true`. Also, set the YB-TServer gflag `ysql_max_write_restart_attempts=0` to disable internal query layer retries on-conflict`

  </tr>
  <tr>
   <td>
   </td>
   <td>

```output
 k | v
---+---
 1 | 1
(1 row)
```

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>

```sql
commit;
```

   </td>
   <td>
   </td>
  </tr>
</tbody>
</table>

#### Rollback of sub-transaction with conflicting write

<table class="no-alter-colors">
<thead>
    <tr>
    <th>
    Client 1
    </th>
    <th>
    Client 2
    </th>
    </tr>
  </thead>
  <tbody>
  <tr>
   <td>

```sql
begin transaction isolation level repeatable read;   
```

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>

```sql
begin transaction isolation level repeatable read;   
```

   </td>
  </tr>
  <tr>
   <td>

```sql
savepoint a;
```

```sql
update test set v=1 where k=1;
```

```output
UPDATE 1
```
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>

```sql
update test set v=1 where k=1;
```

```output
(waits)
```

   </td>
  </tr>
  <tr>
   <td>

```sql
rollback to savepoint a;
```

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>

```output
UPDATE 1
```

```sql
commit;
```

   </td>
  </tr>
  <tr>
   <td>

```sql
commit;
```

   </td>
   <td>
   </td>
  </tr>
</tbody>
</table>

> ### Distributed deadlock detection

Author: nope this is a different section

(waits)
```

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>

```sql
update test set v=6 where k=1;
```

```output
ERROR:  Internal error: Transaction 00da00cd-87fa-431b-9521-253582fb23fe 
was aborted while waiting for locks
```

   </td>
  </tr>
  <tr>
   <td>

```sql
commit;
```

   </td>
   <td>
   </td>
  </tr>
</tbody>
</table>

### Metrics

All metrics are per tablet.

#### Histograms:

1. **wait_queue_pending_time_waiting (ms):** the amount of time a still-waiting transaction has been in the wait queue
2. **wait_queue_finished_waiting_latency (ms):** the amount of time an unblocked transaction spent in the wait queue
3. **wait_queue_blockers_per_waiter:** the number of blockers a waiter is stuck on in the wait queue

#### Counters:

1. **wait_queue_waiters_per_blocker:** the number of waiters stuck on a particular blocker in the wait queue
2. **wait_queue_num_waiters:** the number of waiters stuck on a blocker in the wait queue
3. **wait_queue_num_blockers:** the number of unique blockers tracked in a wait queue

### Limitations

Refer to [#5680](https://github.com/yugabyte/yugabyte-db/issues/5680) for limitations.

## Row-level explicit locking clauses

> The `NOWAIT` clause for row-level explicit locking doesn't make sense in the `Fail-on-Conflict` mode since there is no waiting. It makes sense for the `Wait-on-Conflict` policy but is currently only supported for Read Committed isolation. [#12166](https://github.com/yugabyte/yugabyte-db/issues/12166) will extend support for this in the `Wait-on-Conflict` mode for the other isolation levels.

Reviewer: ```suggestion
The `NOWAIT` clause for row-level explicit locking doesn't apply to the `Fail-on-Conflict` mode as there is no waiting. However, it applies to the `Wait-on-Conflict` policy but is currently supported only for Read Committed isolation. [#12166](https://github.com/yugabyte/yugabyte-db/issues/12166) will extend support for this in the `Wait-on-Conflict` mode for the other isolation levels.
```

> The `SKIP LOCKED` clause is supported in both concurrency control policies and provides a transaction the capability to skip locking without any error when a conflict is detected. However, it isn't supported for Serializable isolation. [#11761](https://github.com/yugabyte/yugabyte-db/issues/5683) tracks support for `SKIP LOCKED` in Serializable isolation.

Reviewer: ```suggestion
The `SKIP LOCKED` clause is supported in both concurrency control policies and provides a transaction with the capability to skip locking without any error when a conflict is detected. However, it isn't supported for Serializable isolation. [#11761](https://github.com/yugabyte/yugabyte-db/issues/5683) tracks support for `SKIP LOCKED` in Serializable isolation.
```

---
title: Concurrency control
headerTitle: Concurrency control
linkTitle: Concurrency control
description: Details about Concurrency control in YSQL
menu:
  preview:
    identifier: architecture-concurrency-control
    parent: architecture-acid-transactions
    weight: 1153
type: docs
---

[Concurrency control](https://en.wikipedia.org/wiki/Concurrency_control) in databases ensures that multiple transactions can execute concurrently while preserving data integrity. Concurrency control is essential for correctness in environments where two or more transactions can access the same data at the same time.

YugabyteDB provides two policies to handle conflicts between concurrent transactions as described in the following sections. 

For information on how row-level explicit locking clauses behave with these concurrency control policies, refer to [Row-level explicit locking clauses](#row-level-explicit-locking-clauses) .

## Fail-on-Conflict

This is the default concurrency control strategy and is applicable for `REPEATABLE READ` and `SERIALIZABLE` isolation levels.

> It is not applicable for [Read Committed](../read-committed/) isolation.

Author: done

In this mode, transactions are assigned random priorities with some exceptions as described in [Transaction Priorities](../transaction-priorities/).

> If a conflict occurs, a transaction with the lower priority is aborted. There are two possibilities when a transaction T1 tries to read or write or lock a row in a mode conflicting with a other concurrent transactions as follows:

Author: done, also kept "other ... transactions" instead of "another ... transaction" since the latter doesn't make sense with the explanations of `wound` and `die` that follow

>    - **Wound:** If T1 has a priority higher than all the other conflicting transactions, T1 will abort them and make progress.

Author: done

Suppose you have a table with some data in it, the following examples describe the wound and die approaches when a conflict occurs.

```sql
create table test (k int primary key, v int);
insert into test values (1, 1);
```

### Wound example

<table class="no-alter-colors">
<thead>
    <tr>
    <th>
    Client 1
    </th>
    <th>
    Client 2
    </th>
    </tr>
  </thead>
  <tbody>
  <tr>
   <td>
   </td>
   <td>

```sql
SET yb_transaction_priority_upper_bound = 0.4;
```
<br>

   </td>
  </tr>
  <tr>
   <td>

```sql
SET yb_transaction_priority_lower_bound = 0.6;
```
<br>

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>

```sql
begin transaction isolation level repeatable read;   
```
<br>

   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>

```sql
select * from test where k=1 for update;
```
<br>

```output
 k | v
---+---
 1 | 1
(1 row)
```

   </td>
  </tr>
  <tr>
   <td>

```sql
begin transaction isolation level repeatable read;   
```
<br>

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>

```sql
select * from test where k=1 for update;
```
<br>

```output
 k | v
---+---
 1 | 1
(1 row)

> ```

Author: done

```

```sql
rollback;
```

   </td>
  </tr>
</tbody>
</table>

### Die example
<table class="no-alter-colors">
<thead>
    <tr>
    <th>
    Client 1
    </th>
    <th>
    Client 2
    </th>
    </tr>
  </thead>
  <tbody>
  <tr>
   <td>
   </td>
   <td>

```sql
SET yb_transaction_priority_lower_bound = 0.6;
```
<br>

   </td>
  </tr>
  <tr>
   <td>

```sql
SET yb_transaction_priority_upper_bound = 0.4;
```
<br>

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>

```sql
begin transaction isolation level repeatable read;   
```
<br>

   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>

```sql
select * from test where k=1 for update;
```
<br>

```output
 k | v
---+---
 1 | 1
(1 row)
```
   </td>
  </tr>
  <tr>
   <td>

```sql
begin transaction isolation level repeatable read;   
```
<br>

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>

```sql
select * from test where k=1 for update;
```
<br>

```output
ERROR:  All transparent retries exhausted. could not serialize
access due to concurrent update
```

```sql
rollback;
```

   </td>
   <td>
   </td>
  </tr>
</tbody>
</table>

### Best-effort internal retries for first statement in a transaction

> Note that the error message `All transparent retries exhausted` in the preceding example is because if the transaction T1, when executing the first statement, finds another concurrent conflicting transaction with equal or higher priority, then T1 will perform a few retries with exponential backoff before giving up in anticipation that the other transaction will be done in some time. The number of retries are configurable by the `ysql_max_write_restart_attempts` YB-TServer gflag and the exponential backoff parameters are the same as the ones described in [Performance tuning](../read-committed/#performance-tuning).

Author: done

> Each retry will work a newer snapshot of the database in anticipation that the conflicts might not occur (if an earlier conflicting transaction T2 has committed with a commit time before the read time of the new snapshot, the conflicts with T2 would essentially be voided since T1 and T2 would no longer be "concurrent").

Author: done

Also note that the retries will not be performed in case the amount of data to be sent from YSQL to the client proxy exceeds the TServer gflag `ysql_output_buffer_size`.

## Wait-on-Conflict [[Beta]](/preview/faq/general/#what-is-the-definition-of-the-beta-feature-tag)

This mode of concurrency control is applicable only for YSQL and provides the same semantics as PostgreSQL.

In this mode, transactions are not assigned priorities. If a conflict occurs when a transaction T1 tries to read/ write/ lock a row in a conflicting mode with a few other concurrent transactions, T1 will **wait** until all conflicting transactions finish by either committing or rolling back. Once all conflicting transactions have finished, T1 will:

1. make progress if the conflicting transactions didn’t commit any permanent modifications that conflict with T1.
2. abort otherwise.

> By default, transactions operate in the priority based `Fail-on-Conflict` mode. This mode can be enabled by setting the YB-TServer gflag `enable_wait_queues=true`, which start in-memory wait queues that provide waiting semantics when conflicts are detected between transactions. A rolling restart is needed for the gflag to take effect.

Author: done

```output
 k | v
---+---
 1 | 1
(1 row)
```

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>

```sql
commit;
```

   </td>
   <td>
   </td>
  </tr>
</tbody>
</table>

#### Rollback of sub-transaction with conflicting write

<table class="no-alter-colors">
<thead>
    <tr>
    <th>
    Client 1
    </th>
    <th>
    Client 2
    </th>
    </tr>
  </thead>
  <tbody>
  <tr>
   <td>

```sql
begin transaction isolation level repeatable read;   
```
<br>

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>

```sql
begin transaction isolation level repeatable read;   
```
<br>

   </td>
  </tr>
  <tr>
   <td>

```sql
savepoint a;
```

```sql
update test set v=1 where k=1;
```

```output
UPDATE 1
```
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>

```sql
update test set v=1 where k=1;
```

```output
(waits)
```

   </td>
  </tr>
  <tr>
   <td>

```sql
rollback to savepoint a;
```

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>

```output
UPDATE 1
```

```sql
commit;
```

   </td>
  </tr>
  <tr>
   <td>

```sql
commit;
```

   </td>
   <td>
   </td>
  </tr>
</tbody>
</table>

### Distributed deadlock detection

> In the Wait-on-Conflict mode, transactions can wait for each other and result in a deadlock. Setting the YB-TServer gflag `enable_deadlock_detection=true` runs a distributed deadlock detection algorithm in the background to detect and break deadlocks. It is always recommended to keep deadlock detection on when `enable_wait_queues=true`, unless the application or workload is such that no deadlocks can occur. A rolling restart is required for the change to take effect.

Author: done

(waits)
```

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>

```sql
update test set v=6 where k=1;
```

```output
ERROR:  Internal error: Transaction 00da00cd-87fa-431b-9521-253582fb23fe 
was aborted while waiting for locks
```

   </td>
  </tr>
  <tr>
   <td>

```sql
commit;
```

   </td>
   <td>
   </td>
  </tr>
</tbody>
</table>

### Metrics

All metrics are per tablet.

#### Histograms:

1. **wait_queue_pending_time_waiting (ms):** the amount of time a still-waiting transaction has been in the wait queue
2. **wait_queue_finished_waiting_latency (ms):** the amount of time an unblocked transaction spent in the wait queue
3. **wait_queue_blockers_per_waiter:** the number of blockers a waiter is stuck on in the wait queue

#### Counters:

1. **wait_queue_waiters_per_blocker:** the number of waiters stuck on a particular blocker in the wait queue
2. **wait_queue_num_waiters:** the number of waiters stuck on a blocker in the wait queue
3. **wait_queue_num_blockers:** the number of unique blockers tracked in a wait queue

### Limitations

Refer to [#5680](https://github.com/yugabyte/yugabyte-db/issues/5680) for limitations.

## Row-level explicit locking clauses

> The `NOWAIT` clause for row-level explicit locking doesn't apply to the `Fail-on-Conflict` mode as there is no waiting. However, it applies to the `Wait-on-Conflict` policy but is currently supported only for Read Committed isolation. [#12166](https://github.com/yugabyte/yugabyte-db/issues/12166) will extend support for this in the `Wait-on-Conflict` mode for the other isolation levels.

Reviewer 2: ```suggestion
The `NOWAIT` clause for row-level explicit locking doesn't apply to the `Fail-on-Conflict` mode as there is no waiting. It does apply to the `Wait-on-Conflict` policy but is currently supported only for Read Committed isolation. [#12166](https://github.com/yugabyte/yugabyte-db/issues/12166) will extend support for this in the `Wait-on-Conflict` mode for the other isolation levels.
```

Author: done

---
title: Concurrency control
headerTitle: Concurrency control
linkTitle: Concurrency control
description: Details about Concurrency control in YSQL
menu:
  preview:
    identifier: architecture-concurrency-control
    parent: architecture-acid-transactions
    weight: 1153
type: docs
---

[Concurrency control](https://en.wikipedia.org/wiki/Concurrency_control) in databases ensures that multiple transactions can execute concurrently while preserving data integrity. Concurrency control is essential for correctness in environments where two or more transactions can access the same data at the same time.

YugabyteDB provides two policies to handle conflicts between concurrent transactions as described in the following sections.

For information on how row-level explicit locking clauses behave with these concurrency control policies, refer to [Row-level explicit locking clauses](#row-level-explicit-locking-clauses) .

## Fail-on-Conflict

This is the default concurrency control strategy and is applicable for `Repeatable Read` and `Serializable` isolation levels.

> It is not applicable for [Read Committed](../read-committed/) isolation.

Author: avoiding this one since it doesn't appear as a hyper link if I add the back ticks

In this mode, transactions are assigned random priorities with some exceptions as described in [Transaction Priorities](../transaction-priorities/).
If a conflict occurs, a transaction with the lower priority is aborted. There are two possibilities when a transaction T1 tries to read, write, or lock a row in a mode conflicting with other concurrent transactions:
   - **Wound:** If T1 has a higher priority than all the other conflicting transactions, T1 will abort them and make progress.
   - **Die:** If any other conflicting transaction has an equal or higher priority than T1, T1 will abort itself.

> Suppose you have a table with some data in it, the following examples describe the wound and die approaches when a conflict occurs.

Author: done

 k | v
---+---
 1 | 1
(1 row)
```
   </td>
  </tr>
  <tr>
   <td>

```sql
begin transaction isolation level repeatable read;
```
<br>

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>

```sql
select * from test where k=1 for update;
```
<br>

```output
ERROR:  All transparent retries exhausted. could not serialize
access due to concurrent update
```

```sql
rollback;
```

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>

```sql
commit;
```

   </td>
  </tr>
</tbody>
</table>

### Best-effort internal retries for first statement in a transaction

Note that we see the error message `All transparent retries exhausted` in the preceding example because if the transaction T1, when executing the first statement, finds another concurrent conflicting transaction with equal or higher priority, then T1 will perform a few retries with exponential backoff before giving up in anticipation that the other transaction will be done in some time. The number of retries are configurable by the `ysql_max_write_restart_attempts` YB-TServer gflag and the exponential backoff parameters are the same as the ones described in [Performance tuning](../read-committed/#performance-tuning).

> Each retry will use a newer snapshot of the database in anticipation that the conflicts might not occur (if an earlier conflicting transaction T2 has committed with a commit time before the read time of the new snapshot, the conflicts with T2 would essentially be voided since T1 and T2 would no longer be "concurrent").

Author: modified in a slightly different way since this is not an example but the generic reasoning that applies always

Note that the retries will not be performed in case the amount of data to be sent from YSQL to the client proxy exceeds the TServer gflag `ysql_output_buffer_size`.

> ## Wait-on-Conflict [[Beta]](/preview/faq/general/#what-is-the-definition-of-the-beta-feature-tag) {#wait-on-conflict}

Reviewer 3: The `{#wait-on-conflict}` is a manually-added anchor. Don't remove it, and links to this heading will stay working even after we remove the beta tag.

This mode of concurrency control is applicable only for YSQL and provides the same semantics as PostgreSQL.

In this mode, transactions are not assigned priorities. If a conflict occurs when a transaction T1 tries to read, write, or lock a row in a conflicting mode with a few other concurrent transactions, T1 will **wait** until all conflicting transactions finish by either committing or rolling back. Once all conflicting transactions have finished, T1 will:

1. make progress if the conflicting transactions didn’t commit any permanent modifications that conflict with T1.

> 2. abort otherwise.

Author: I think it is confusing in this context, better to keep it as a separate note section below

> By default, transactions operate in the priority based `Fail-on-Conflict` mode. This mode can be enabled by setting the YB-TServer gflag `enable_wait_queues=true`, which will enable use of in-memory wait queues that provide waiting semantics when conflicts are detected between transactions. A rolling restart is needed for the gflag to take effect.

Author: done

   1. the app will have to [rely on statement timeouts to avoid deadlocks](../read-committed/#avoid-deadlocks-in-read-committed-transactions).
   1. there can be unfairness and high P99 latencies during contention due to the retry-backoff mechanism.
{{</note >}}

{{< note title="Best-effort internal retries also apply to Wait-on-Conflict policy" >}}

The best-effort internal retries described in Fail-on-Conflict apply to Wait-on-Conflict policy as well. YugabyteDB provides this additional enhancement which is not supported by PostgreSQL.

After a transaction T1 (that was waiting for other transactions) unblocks, it could be the case that some conflicting modifications were committed to the database. In this case, T1 has to abort. However, if its still the first statement that was being executed in T1, best-effort internal retries using a later snapshot of the database will be performed to possibly make progress.

{{</note >}}

### Examples

The following examples describe different use cases detailing the Wait-on-Conflict behavior.

> Note that the examples require you to set the YB-TServer gflag `enable_wait_queues=true`. Also, set the YB-TServer gflag `ysql_max_write_restart_attempts=0` to disable internal query layer retries on-conflict. A restart is necessary for these flags to take effect.

Author: done

```
<br>

   </td>
   <td>

```sql
begin transaction isolation level repeatable read;
```
<br>

   </td>
  </tr>
  <tr>
   <td>

```sql
select * from test where k=1 for share;
```
<br>

```output
 k | v
---+---
 1 | 1
(1 row)
```

   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>

```sql
select * from test where k=1 for update;
```
<br>

```
(waits for T1 to end...)
```

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>
   </td>
   <td>

```sql
select * from test where k=1 for share;
```
<br>

```output
 k | v
---+---
 1 | 1
(1 row)
```

(Doesn't wait for T2 even though it conflicts with the explicit row-level lock T2 is waiting for)
   </td>
  </tr>
  <tr>
   <td>

```sql
commit;
```

   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>
   </td>
   <td>

```sql
commit;
```

   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>

```output
 k | v
---+---
 1 | 1
(1 row)
```

   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>
   </td>
   <td>

```sql
commit;
```

   </td>
   <td>
   </td>
  </tr>
</tbody>
</table>

#### Rollback of sub-transaction with conflicting write

> 

Reviewer 2: ```suggestion

If a transaction T1 was blocked on some operation from a separate transaction T2, and that blocking operation was part of a subtransaction which is rolled back, then T1 may proceed:
```

Author: done
